keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,432 @@
+## Drawbacks
+
+The main drawback associated with the approach leveraging csi-proxy.exe is that the life cycle of that binary as well as logs will need to be managed out-of-band from Kubernetes. However, cluster administrators need to maintain and manage life cycle and logs of other core binaries like kubeproxy.exe, kubelet.exe, dockerd.exe and containerd.exe (in the future). Therefore csi-proxy.exe will be one additional binary that will need to be treated in a similar way.
+
+The API versioning scheme described above, as of now, will maintain backward compatibility indefinitely. This requires the scope of csi-proxy.exe to be limited to a very scoped down fundamental set of operations. Maintainers therefore will need to be very cautious when accepting suggestions for new APIs and enhancements. This may slow progress at times.
+
+There may ultimately be certain operations that csi-proxy.exe cannot support in an elegant fashion and require the plugin author targeting Windows nodes to seek one of the alternatives described below. There may also be a need to support volumes that do not use standard block or file protocols. In such scenarios, an extra payload (in the form of a binary, kernel driver and service) may need to be dropped on the host and maintained out-of-band from Kubernetes. This KEP and maintainers should ensure such instances are as limited as possible.
+
+## Alternatives
+
+There are alternative approaches to the CSIProxyService API as well as the overall csi-proxy mechanism described in this KEP. These alternatives are enumerated below.
+
+### API Alternatives
+
+The CSIProxyService API will be a defined set of operations that will need to expand over time as new CSI APIs are introduced that require new operations on every node as well as desire for richer operations by CSI plugin authors. Unfortunately this comes with a maintenance burden associated with tracking and graduating new RPCs across versions.
+
+An alternative approach that simplifies the above involves exposing a single Exec interface in CSIProxyService that supports passing an arbitrary set of parameters to arbitrary executables and powershell cmdlets on the Windows host and collecting and returning the stdout and stderr back to the invoking containerized process. Since all the currently enumerated operations in the CSIProxyService API can be driven through the generic Exec interface, the set of desired privileged operations necessary becomes a decision for plugin authors rather than maintainers of csi-proxy. The main drawback of this highly flexible mechanism is that it drastically increases the potential for abuse in the host by 3rd party plugin authors. The effect of bugs or vulnerabilities in individual plugins and ability to exploit them to take control of the host becomes critical. Depending on the adoption of csi-proxy in the Alpha and Beta phases and the need for specialized privileged operations, we may consider adding a generic Exec interface in the future.
```

Comment:
If a driver is compromised, would exec even pose an even greater risk than the api surface in the main design? Looks like you could expose arbitrary directories to arbitrary places effectively giving you control over the entire system (maybe root through some finite number of steps). Not a security or windows expert though :)

Reply:
Agree that with a compromised plugin, a lot of damage can be done through the privileged interface. I will clarify that the security concern should not be a reason to pre-clue an `Exec` interface if we do need one.